### ISSUE TITLE
"Can't create the event."

### ISSUE THREAD

**Issue by PatrykRzadzinski** [FabianHenneke/ics2gcal#1]
Once I right-click an .ics on a webpage (nextcloud), a message appears about an attempt to add it to my calendar, but follows with another one saying "Can't create the event.".

**Comment by FabianHenneke** [comment]
Would it be possible for you to provide me with an example .ics that causes this behavior? You could download one and strip it of all personal information using any text editor.

**Comment by faustov** [comment]
[ics.zip](https://github.com/FabianHenneke/ics2gcal/files/2610302/ics.zip)
Certainly - attached. My editor could potentially replace windows crlf into unix, not sure if it matters.

**Comment by FabianHenneke** [comment]
Creating the event fails with the following 400 response from the Google Calendar API:
```
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "invalid",
    "message": "Invalid time zone definition for start time."
   },
   {
    "domain": "global",
    "reason": "required",
    "message": "Missing time zone definition for start time."
   },
   {
    "domain": "global",
    "reason": "invalid",
    "message": "Invalid time zone definition for end time."
   },
   {
    "domain": "global",
    "reason": "required",
    "message": "Missing time zone definition for end time."
   }
  ],
  "code": 400,
  "message": "Invalid time zone definition for start time."
 }
}
```

This is caused by the following two lines in the .ics:

```
DTSTART;TZID=W. Europe Standard Time:20181130T113000
DTEND;TZID=W. Europe Standard Time:20181130T120000
```

The .ics uses a non-standard name of a time zone which is not handled by Google Calendar. Since ICS to GCal is not meant to support more calendar functionality than Google Calendar, I am labelling this as wontfix. 

Thank you very much for taking your time to report this though. If you think that this is an issue I should fix somehow, please let me know. Closing this for now.

**Comment by PatrykRzadzinski** [comment]
First of all, thanks for looking into this! Unfortunatelly, as the only vendor of a solution to load .ics items I get over email via chrome to my google calendar, I cannot let you off the hook so easily ;)

As you can see this .ics has been generated by MS Exchange, which I'm sure you can appreciate is not a niche vendor, even when it comes to calendar events only...

Considering the name of the timezone is completely meaningless, why not add a condition to ignore it if it whenever "non-standard" is detected?

**Comment by FabianHenneke** [comment]
The only "simple" solution I see right now would be to catch the error message returned by the Google Calendar API and retry without explicit timezones. This sounds like an ugly hack to me though, both because the error strings returned by the API may change in the future (they are not meant to be machine-readable) and since it silently assumes that the user only adds events specified in their current local time zone.

Alternatively, one could provide a mapping of MS time zones to IANA time zone names, maybe there is even a library for this. I will keep this at the back of my mind and see whether I can do this at some point. In the meantime, I will leave this open for people to make alternative suggestions or even submit a PR that ensures MS exchange compatibility.

**Comment by PatrykRzadzinski** [comment]
Thanks Fabian. Dammit, I got another one. Seems vast majority originate from MS products. There are indeed mapping tables available in XML, but I wonder, why not ignore TZID since TZOFFSETFROM and
TZOFFSETTO are provided in each .ics file?

**Comment by PatrykRzadzinski** [comment]
I got an invite straight out of G-suite today and importing it to my calendar ended with the same result. Maybe the root cause is different? Attaching this ics.
[gsuite.ics.zip](https://github.com/FabianHenneke/ics2gcal/files/2652353/gsuite.ics.zip)

**Comment by FabianHenneke** [comment]
> Thanks Fabian. Dammit, I got another one. Seems vast majority originate from MS products. There are indeed mapping tables available in XML, but I wonder, why not ignore TZID since TZOFFSETFROM and
> TZOFFSETTO are provided in each .ics file?

I will check whether the .ics library I use can convert custom timezones into a format that the Google Calendar accepts.

> I got an invite straight out of G-suite today and importing it to my calendar ended with the same result. Maybe the root cause is different? Attaching this ics.
> [gsuite.ics.zip](https://github.com/FabianHenneke/ics2gcal/files/2652353/gsuite.ics.zip)

If I serve this file from a local web server and use ICS to GCal on a link to it, the event is imported successfully. I suspect that something is wrong with the GSuite link, maybe it uses a JS redirect or something else that a simple `fetch()` to the its `href` can't handle.

**Comment by PatrykRzadzinski** [comment]
Here's how it looks:
https://nextcloud.com/apps/rainloop/app/?/Raw/&q[]=/_qNNNo-seEOrVAC-ZVU6VQhLJ0dIxyT9IC34TBCHJJflKewjFMWv0f5L5hS7LGTg3F6Zp9_75GUquajEec8qT40pv_apGpPOtT03N__kUl4Dasvh6Ez9LYsjWFYl5JamN3cEWpIobYvMwvC-m6yzr1O4iqOGuM_vf2xhKoOgKXBdls-wdHWYgEaFGmdvWy0aa62b9hPkRZdQQZQkk2Avl3pgGQuH7ooHqAqHowzrui20-wf1lOv6vHkhSPYLn9SRbrG9VqOIF69V0Dwh6aJnHsIr3D6Y2sl_fdNaA7JwZJrhZ7jvZaN0-HSaygZ14iaXK5fODVJRS45bGGoF4/ViewAsPlain/&q[]=/3EU8E546RPechd1bMDCUoIDU0TRz-Qs2M15gQGa_kaopd1meGdGnjj6DxPhw7NRoTPehHV1lOSlzxLq_ubAKqRwmTecBe8guZnVX6VIjNgs395YBA0Z1gC8bRvLST4PY-QB6OvvU-bO5X5P8OkREmsi-3aO3QlkOl2yzC9cAhj0wHLBY5BUBK71WlTA5Eqc5IOaETVA9DPRu4HfawaCIh4QmOQMrBfFrcvETypg4xW8v2Sh8bDwufRKVcLp7xyNgOaszmDxA3rfAsMmQaVEZCKPW9SDPM8rC4aWQ2s1pTo10Fa_9-kl67Bp4C1vxq2XgcRNwjmEThqazzOR3

When I click this link, the .ics file gets downloaded by Chrome.

Here's how it looks
![image](https://user-images.githubusercontent.com/5629403/49722628-78009280-fc65-11e8-889f-ff5dcbdf2eb1.png)

**Comment by FabianHenneke** [comment]
That link gives me a 404 Page not found. Please check whether it works for you in Incognito/Private Mode.

**Comment by PatrykRzadzinski** [comment]
Following the link in incognito mode opens a blank page and nothing more. Should I report an issue to rainloop?

**Comment by FabianHenneke** [comment]
You could try that, but I don't know enough about nextcloud or rainloop to assess whether this is actually a bug or not. It's very likely not an issue with ICS to GCal though. 

I will look into the timezone issue when I have time. If you encounter other .ics files in the wild which you are reasonably sure are being handled incorrectly by ICS to GCal, please open another issue.

**Comment by PatrykRzadzinski** [comment]
Sounds good, thanks!

**Comment by FabianHenneke** [comment]
Microsoft Exchange uses the Windows Time Zone Indices, whereas the Google Calendar does only support the standardized IANA identifiers. Since there is no way to import a custom timezone into Google Calendar, it would be necessary to implement a mapping between the two systems. The authoritative source would be `https://docs.microsoft.com/en-us/rest/api/maps/timezone/gettimezonewindowstoiana`, but this requires a commercial API key.
Another option would be to hardcode the list and hope that I get notified when time zones change. Would you say that hardcoding the timezone mapping would be an acceptable solution?

**Comment by PatrykRzadzinski** [comment]
I can't think of anything better. To my understanding, timezones don't change too often anymore.

**Comment by FabianHenneke** [comment]
I have added a conversion from Windows to IANA time zones in https://github.com/FabianHenneke/ics2gcal/commit/ae6ef603265a358bf8604c1a5297e3f1383aa230. Could you confirm that this works with your Exchange-generated .ics files? If it does, I would push it to the web store. Thanks again for bringing this to my attention.

**Comment by PatrykRzadzinski** [comment]
Works like a charm ;) Thank you!

**Comment by FabianHenneke** [comment]
Included in 0.6.0.
